src/System.Management.Automation/engine/hostifaces/PSTask.cs

```diff
@@ -842,6 +897,64 @@ private void CheckForComplete()
             }
         }
 
+        private Runspace GetRunspace(int taskId)
+        {
+            var runspaceName = string.Format(CultureInfo.InvariantCulture, "{0}:{1}", PSTask.RunspaceName, taskId);
+
+            if (_useRunspacePool && _runspacePool.TryDequeue(out Runspace runspace))
+            {
+                if (runspace.RunspaceStateInfo.State == RunspaceState.Opened &&
+                    runspace.RunspaceAvailability == RunspaceAvailability.Available)
+                {
+                    try
+                    {
+                        runspace.ResetRunspaceState();
+                        runspace.Name = runspaceName;
+                        return runspace;
+                    }
+                    catch
+                    {
+                        // If the runspace cannot be reset for any reason, remove it.
+                    }
+                }
+
+                RemoveActiveRunspace(runspace);
```

Comment:
Do we want to try dequeue again in case the runspace is broken?

Reply:
I feel it is better to replace a broken runspace. The task pool will ensure maximum number of threads/runspaces are used.

Reply:
- When the pipeline objects are not coming in that fast and the parallel script block takes a small time to finish, it's likely there is no need to replace the broken runspace -- rest of the available runspaces can handle all incoming tasks.
- When tasks are coming in fast, the broken runspace will be replaced when all available runspace are used up.

So it's possible try dequeuing again until the queue is empty can have better performance. But I guess the chance of having a broken runspace in the queue is low, so the difference won't be noticeable.

Reply:
I understand.  But this is a tail off scenario.  Another strategy is steady state, maintaining maximum pool tasks.  But I agree that a broken runspace is unlikely (except for a pathological script which would probably crash the process anyway).  At this point I would like to see actual perf scenarios (such as the great work precipitating this change) before making more changes.  I have been thinking about pooling threads as well, but there are problems with that and it is not clear what perf improvement would result.